src/checker.py

```diff
@@ -173,12 +171,26 @@ def _collect_module_data(self, module_path: str, module: t.Union[str, t.Dict]):
         for child_module in child_modules:
             self._collect_module_data(module_path=module_path, module=child_module)
 
-        sources = module.get("sources", [])
-
         manifest_datas = self._external_data.setdefault(module_path, [])
+
+        sources: t.List[t.Dict] = []
+        source_paths: t.List[str] = []
+        for source in module.get("sources", []):
+            # Find external sources (path strings) first
+            if isinstance(source, str):
+                if _external_source_filter(module_path, source):
+                    source_paths.append(source)
+                continue
+            assert isinstance(source, dict)
+            # Next, collect only unique sources
+            # NOTE here we rely on ruamel.yaml to make YAML aliases into
+            # pointers to the same dict object ridden from YAML anchor
+            if not any(d.source is source for d in manifest_datas):
```

Comment:
O(n²) in the number of sources. Maybe it's fine because 'n' is small.

But I can't quite see how this will work – `manifest_datas` is not updated until after this loop, so if there are aliased sources in the same file, how will this catch them? Oh, because this loop is for sources in a particular module, and `manifest_datas` is for the entire file. So we would not catch duplicates in the case of the same source being aliased in the same module, but that will not happen.

Any idea how to test this? My suggestion would be:

- Manifest using YAML anchors for a source
- Mock checker in the test case which records how many times it has been invoked
- Assert that it is invoked exactly once

Reply:
>O(n²) in the number of sources. Maybe it's fine because 'n' is small.

I believe `X is Y` compares only pointers, not actual objects, so this should be fast enough.

>we would not catch duplicates in the case of the same source being aliased in the same module, but that will not happen

Ah, good point. This is unlikely to happen in practice, yes, but I think this should be fixed anyway. I was thinking of splitting "collect data from source" logic to separate method, it would be called per-source, so it'll probably will fix this.

>Any idea how to test this?

Just run the checker against manifest with, say, 2 anchors with 1 alias each, and very that we collected exactly 2 data items?

Reply:
@wjt As the easiest way to test this, I've added an alias to our enrypoint-testing manifest (the test verifies amount of outdated sources), and rearranged this code block in a way that we can see if both condition outcomes are covered.
Do you think this is sufficient?

Reply:
Now we also directly check amount of data collected in `test_checker.py`